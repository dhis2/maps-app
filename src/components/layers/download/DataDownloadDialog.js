import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import i18n from '@dhis2/d2-i18n';
import { Modal, ModalTitle, ModalContent, ModalActions } from '@dhis2/ui';
import { EVENT_LAYER } from '../../../constants/layers';

import {
    META_DATA_FORMAT_ID,
    META_DATA_FORMAT_NAME,
    META_DATA_FORMAT_CODE,
} from '../../../util/geojson';

import {
    closeDataDownloadDialog,
    startDataDownload,
} from '../../../actions/dataDownload';
import DataDownloadDialogContent from './DataDownloadDialogContent';
import DataDownloadDialogActions from './DataDownloadDialogActions';

const formatOptionsFlat = [
    META_DATA_FORMAT_ID,
    META_DATA_FORMAT_CODE,
    META_DATA_FORMAT_NAME,
];
const formatOptions = formatOptionsFlat.map((name, i) => ({
    id: i + 1,
    name,
}));

export class DataDownloadDialog extends Component {
    static propTypes = {
        open: PropTypes.bool.isRequired,
        downloading: PropTypes.bool.isRequired,
        error: PropTypes.string,
        layer: PropTypes.object,
        aggregations: PropTypes.object,
        startDownload: PropTypes.func.isRequired,
        closeDialog: PropTypes.func.isRequired,
    };

    state = {
        selectedFormatOption: 2,
        humanReadableChecked: true,
    };

    onChangeFormatOption = newValue => {
        this.setState({ selectedFormatOption: newValue.id - 1 });
    };
    onCheckHumanReadable = isChecked => {
        this.setState({ humanReadableChecked: isChecked });
    };

    onStartDownload = () => {
        const { layer, aggregations } = this.props;
        const {
            selectedFormatOption,
            humanReadableChecked: humanReadableKeys,
        } = this.state;
        const format = formatOptionsFlat[selectedFormatOption];

        this.props.startDownload({
            layer,
            aggregations,
            format,
            humanReadableKeys,
        });
    };

    render() {
        const { open, layer, downloading, error, closeDialog } = this.props;

        if (!open || !layer) {
            return null;
        }

        const layerType = layer.layer,
            isEventLayer = layerType === EVENT_LAYER;

        return (
            <Modal position="middle" onClose={closeDialog}>
                <ModalTitle>{i18n.t('Download Layer Data')}</ModalTitle>
                <ModalContent>
                    <DataDownloadDialogContent
                        isEventLayer={isEventLayer}
                        error={error}
                        layerName={layer.name}
                        formatOptions={formatOptions}
                        selectedFormatOption={
                            this.state.selectedFormatOption + 1
                        }
                        humanReadableChecked={this.state.humanReadableChecked}
                        onChangeFormatOption={this.onChangeFormatOption}
                        onCheckHumanReadable={this.onCheckHumanReadable}
                    />
                </ModalContent>
                <ModalActions>
                    <DataDownloadDialogActions
                        downloading={downloading}
                        onStartClick={this.onStartDownload}
                        onCancelClick={closeDialog}
                    />
                </ModalActions>
            </Modal>
        );
    }
}

const mapStateToProps = ({ dataDownload, map, aggregations = {} }) => {
    const { layerid, dialogOpen: open, downloading, error } = dataDownload;
    const layer = map.mapViews.find(l => l.id === layerid);

    if (open && !layer) {
        // eslint-disable-next-line
        console.error(
            'Tried to open data download dialog without specifying a source layer!'
        );
    }

    return {
        open,
        layer,
        downloading,
        error,
        aggregations: aggregations[layerid],
    };
};

export default connect(mapStateToProps, {
    closeDialog: closeDataDownloadDialog,
    startDownload: startDataDownload,
})(DataDownloadDialog);
