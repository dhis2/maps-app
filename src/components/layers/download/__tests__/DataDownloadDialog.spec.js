import React from 'react';
import { shallow } from 'enzyme';
import { DataDownloadDialog } from '../DataDownloadDialog';

describe('DataDownloadDialogContent', () => {
    const formatOptions = [
        { id: 1, name: 'Test Format' },
        { id: 2, name: 'No Format' },
        { id: 3, name: 'Some Format' },
    ];
    const renderComponent = props =>
        shallow(
            <DataDownloadDialog
                classes={{}}
                open={false}
                downloading={false}
                error={null}
                layer={null}
                startDownload={() => null}
                closeDialog={() => null}
                {...props}
            />
        );

    const dummyEventLayer = { layer: 'event' };
    const dummyThematicLayer = { layer: 'thematic' };

    it('Should render null when not open', () => {
        const wrapper = renderComponent();
        expect(wrapper.isEmptyRender()).toBe(true);
    });

    it('Should render null when open but layer is null', () => {
        const wrapper = renderComponent({
            open: true,
        });
        expect(wrapper.isEmptyRender()).toBe(true);
    });

    it('Should render a dialog when open', () => {
        const wrapper = renderComponent({
            open: true,
            layer: dummyEventLayer,
        });
        expect(wrapper.isEmptyRender()).toBe(false);
        expect(wrapper.find('WithStyles(Dialog)').length).toBe(1);
        expect(
            wrapper.find('WithStyles(DataDownloadDialogContent)').length
        ).toBe(1);
        expect(
            wrapper.find('WithStyles(DataDownloadDialogActions)').length
        ).toBe(1);
        expect(
            wrapper
                .find('WithStyles(DataDownloadDialogActions)')
                .prop('downloading')
        ).toBe(false);
        expect(
            wrapper
                .find('WithStyles(DataDownloadDialogContent)')
                .prop('isEventLayer')
        ).toBe(true);
    });

    it('Should detect isEventLayer correctly', () => {
        const wrapper = renderComponent({
            open: true,
            layer: dummyThematicLayer,
        });
        expect(wrapper.isEmptyRender()).toBe(false);
        expect(
            wrapper
                .find('WithStyles(DataDownloadDialogContent)')
                .prop('isEventLayer')
        ).toBe(false);
    });
    it('Should default state to selectedFormatOption=2 and humanReadableChecked=true', () => {
        const wrapper = renderComponent({
            open: true,
            layer: dummyEventLayer,
        });
        expect(wrapper.isEmptyRender()).toBe(false);
        expect(wrapper.state().selectedFormatOption).toBe(2);
        expect(
            wrapper
                .find('WithStyles(DataDownloadDialogContent)')
                .prop('selectedFormatOption')
        ).toBe(3); // 1-indexed
        expect(wrapper.state().humanReadableChecked).toBe(true);
        expect(
            wrapper
                .find('WithStyles(DataDownloadDialogContent)')
                .prop('humanReadableChecked')
        ).toBe(true);
    });
    it('Should update humanReadableChecked when calling onCheckHumanReadable', () => {
        const wrapper = renderComponent({
            open: true,
            layer: dummyEventLayer,
        });
        expect(wrapper.isEmptyRender()).toBe(false);
        expect(wrapper.state().humanReadableChecked).toBe(true);
        wrapper.instance().onCheckHumanReadable(false);
        expect(wrapper.state().humanReadableChecked).toBe(false);
    });
    it('Should update selectedFormatOption when calling onChangeFormatOption', () => {
        const wrapper = renderComponent({
            open: true,
            layer: dummyEventLayer,
        });
        expect(wrapper.isEmptyRender()).toBe(false);
        expect(wrapper.state().selectedFormatOption).toBe(2);
        wrapper.instance().onChangeFormatOption({ id: 1 }); // IDs are 1-indexed, selectedFormatOption is 0-indexed
        expect(wrapper.state().selectedFormatOption).toBe(0); // 1 - 1 = 0
    });
    it('Should pass downloading to Actions', () => {
        const wrapper = renderComponent({
            open: true,
            layer: dummyEventLayer,
            downloading: true,
        });
        expect(wrapper.find('WithStyles(DataDownloadDialogActions)').prop('downloading')).toBe(true);
    });
});
