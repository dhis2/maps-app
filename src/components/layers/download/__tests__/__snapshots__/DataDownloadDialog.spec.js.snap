// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataDownloadDialogContent Should render a dialog when open 1`] = `
<div>
  <div
    class="ui-Modal"
  >
    <div
      class="ui-ModalTitle"
    >
      Download Layer Data
    </div>
    <div
      class="ui-ModalContent"
    >
      <div
        class="contentDiv"
      >
        Downloading GeoJSON data for "Thematic layer"
      </div>
      <div
        class="help"
      >
        <svg
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 1a7 7 0 110 14A7 7 0 018 1zm0 1a6 6 0 100 12A6 6 0 008 2zm.5 5v4H10v1H6v-1h1.5V7zM8 4a.75.75 0 110 1.5A.75.75 0 018 4z"
            fill="currentColor"
          />
        </svg>
        <div
          class="content"
        >
          GeoJSON is supported by most GIS software, including QGIS and ArcGIS Desktop.
        </div>
      </div>
    </div>
    <div
      class="ui-ModalActions"
    >
      <div
        class="ui-ButtonStrip"
      >
        <div
          class="ui-Button"
        >
          Cancel
        </div>
        <div
          class="ui-Button"
        >
          Download
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DataDownloadDialogContent does not render when not open 1`] = `<div />`;

exports[`DataDownloadDialogContent renders for an event layer 1`] = `
<div>
  <div
    class="ui-Modal"
  >
    <div
      class="ui-ModalTitle"
    >
      Download Layer Data
    </div>
    <div
      class="ui-ModalContent"
    >
      <div
        class="contentDiv"
      >
        Downloading GeoJSON data for "Event layer"
      </div>
      <div
        class="help"
      >
        <svg
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 1a7 7 0 110 14A7 7 0 018 1zm0 1a6 6 0 100 12A6 6 0 008 2zm.5 5v4H10v1H6v-1h1.5V7zM8 4a.75.75 0 110 1.5A.75.75 0 018 4z"
            fill="currentColor"
          />
        </svg>
        <div
          class="content"
        >
          GeoJSON is supported by most GIS software, including QGIS and ArcGIS Desktop.
        </div>
      </div>
      <div
        class="inputContainer"
      >
        <div
          class="headingDiv"
        >
          GeoJSON Properties:
        </div>
        <div
          class="selectField"
        >
          <div
            class="inputField"
          >
            <div
              class="jsx-1141087211 "
              data-test="dhis2-uiwidgets-singleselectfield"
            >
              <label
                class="jsx-2904627559 "
                data-test="dhis2-uiwidgets-singleselectfield-label"
              >
                <span
                  class="jsx-2904627559"
                >
                  ID Format
                </span>
              </label>
              <div
                class="jsx-3651602664 "
                data-test="dhis2-uiwidgets-singleselectfield-content"
              >
                <div
                  class="jsx-3415779570 "
                  data-test="dhis2-uicore-box"
                >
                  <div
                    class="jsx-114080822 root"
                    data-test="dhis2-uicore-singleselect"
                  >
                    <div
                      class="jsx-114080822 root-input"
                    >
                      <div
                        data-test="dhis2-uicore-select"
                      >
                        <div
                          class="jsx-1462515063 root dense"
                          data-test="dhis2-uicore-select-input"
                          tabindex="0"
                        >
                          <div
                            class="jsx-1462515063 root-children"
                          >
                            <div
                              class="jsx-2240447017 jsx-4123978364 root"
                            >
                              <div
                                class="jsx-2240447017 jsx-4123978364 root-input"
                              >
                                <div
                                  class="jsx-1323585369 root"
                                >
                                  Name
                                </div>
                              </div>
                            </div>
                          </div>
                          <div
                            class="jsx-1462515063 root-right"
                          >
                            <svg
                              height="16"
                              viewBox="0 0 16 16"
                              width="16"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M4.146 6.146a.5.5 0 01.638-.057l.07.057L8 9.293l3.146-3.147a.5.5 0 01.638-.057l.07.057a.5.5 0 01.057.638l-.057.07-3.5 3.5a.5.5 0 01-.638.057l-.07-.057-3.5-3.5a.5.5 0 010-.708z"
                                fill="currentColor"
                              />
                            </svg>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="checkbox checkboxRoot"
        >
          <label
            class="jsx-4249355495 dense"
            data-test="dhis2-uicore-checkbox"
          >
            <input
              checked=""
              class="jsx-4249355495"
              type="checkbox"
              value=""
            />
            <div
              class="jsx-4249355495 icon dense"
            >
              <svg
                class="jsx-4069423049 jsx-3997907426 checked"
                viewBox="0 0 14 14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  class="jsx-4069423049 jsx-3997907426 background"
                  height="14"
                  rx="2"
                  width="14"
                  x="0"
                  y="0"
                />
                <path
                  class="jsx-4069423049 jsx-3997907426 border"
                  d="M2,0 L12,0 C13.1045695,0 14,0.8954305 14,2 L14,12 C14,13.1045695 13.1045695,14 12,14 L2,14 C0.8954305,14 0,13.1045695 0,12 L0,2 C0,0.8954305 0.8954305,0 2,0 Z M2,1 C1.44771525,1 1,1.44771525 1,2 L1,12 C1,12.5522847 1.44771525,13 2,13 L12,13 C12.5522847,13 13,12.5522847 13,12 L13,2 C13,1.44771525 12.5522847,1 12,1 L2,1 Z"
                />
                <path
                  class="jsx-4069423049 jsx-3997907426 checkmark"
                  d="M10.3520005,3.30015877 L5.54268293,8.03252726 L3.64634146,6.16652726 L2.28668615,7.5044281 L5.54310367,10.7010587 L11.7145993,4.628307 L10.3520005,3.30015877 Z M5.54268293,9.43547274 L5.61193184,9.36733181 L5.54226219,9.29894127 L5.47309681,9.367 L5.54268293,9.43547274 Z M10.6496329,4.41023392 L10.6495057,4.35652297 L10.5648412,4.35672356 L10.2854007,4.631693 L10.3553166,4.69984123 L10.6496329,4.41023392 Z M3.42891348,7.22439304 L3.35030905,7.22439304 L3.35034135,7.27820863 L3.64634146,7.56947274 L3.71331385,7.5035719 L3.42891348,7.22439304 Z"
                />
                <rect
                  class="jsx-4069423049 jsx-3997907426 indeterminate"
                  height="2"
                  width="8"
                  x="3"
                  y="6"
                />
              </svg>
            </div>
            Use human-readable keys
          </label>
        </div>
      </div>
    </div>
    <div
      class="ui-ModalActions"
    >
      <div
        class="ui-ButtonStrip"
      >
        <div
          class="ui-Button"
        >
          Cancel
        </div>
        <div
          class="ui-Button"
        >
          Download
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DataDownloadDialogContent renders for event layer with downloading = true 1`] = `
<div>
  <div
    class="ui-Modal"
  >
    <div
      class="ui-ModalTitle"
    >
      Download Layer Data
    </div>
    <div
      class="ui-ModalContent"
    >
      <div
        class="contentDiv"
      >
        Downloading GeoJSON data for "Event layer"
      </div>
      <div
        class="help"
      >
        <svg
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 1a7 7 0 110 14A7 7 0 018 1zm0 1a6 6 0 100 12A6 6 0 008 2zm.5 5v4H10v1H6v-1h1.5V7zM8 4a.75.75 0 110 1.5A.75.75 0 018 4z"
            fill="currentColor"
          />
        </svg>
        <div
          class="content"
        >
          GeoJSON is supported by most GIS software, including QGIS and ArcGIS Desktop.
        </div>
      </div>
      <div
        class="inputContainer"
      >
        <div
          class="headingDiv"
        >
          GeoJSON Properties:
        </div>
        <div
          class="selectField"
        >
          <div
            class="inputField"
          >
            <div
              class="jsx-1141087211 "
              data-test="dhis2-uiwidgets-singleselectfield"
            >
              <label
                class="jsx-2904627559 "
                data-test="dhis2-uiwidgets-singleselectfield-label"
              >
                <span
                  class="jsx-2904627559"
                >
                  ID Format
                </span>
              </label>
              <div
                class="jsx-3651602664 "
                data-test="dhis2-uiwidgets-singleselectfield-content"
              >
                <div
                  class="jsx-3415779570 "
                  data-test="dhis2-uicore-box"
                >
                  <div
                    class="jsx-114080822 root"
                    data-test="dhis2-uicore-singleselect"
                  >
                    <div
                      class="jsx-114080822 root-input"
                    >
                      <div
                        data-test="dhis2-uicore-select"
                      >
                        <div
                          class="jsx-1462515063 root dense"
                          data-test="dhis2-uicore-select-input"
                          tabindex="0"
                        >
                          <div
                            class="jsx-1462515063 root-children"
                          >
                            <div
                              class="jsx-2240447017 jsx-4123978364 root"
                            >
                              <div
                                class="jsx-2240447017 jsx-4123978364 root-input"
                              >
                                <div
                                  class="jsx-1323585369 root"
                                >
                                  Name
                                </div>
                              </div>
                            </div>
                          </div>
                          <div
                            class="jsx-1462515063 root-right"
                          >
                            <svg
                              height="16"
                              viewBox="0 0 16 16"
                              width="16"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M4.146 6.146a.5.5 0 01.638-.057l.07.057L8 9.293l3.146-3.147a.5.5 0 01.638-.057l.07.057a.5.5 0 01.057.638l-.057.07-3.5 3.5a.5.5 0 01-.638.057l-.07-.057-3.5-3.5a.5.5 0 010-.708z"
                                fill="currentColor"
                              />
                            </svg>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="checkbox checkboxRoot"
        >
          <label
            class="jsx-4249355495 dense"
            data-test="dhis2-uicore-checkbox"
          >
            <input
              checked=""
              class="jsx-4249355495"
              type="checkbox"
              value=""
            />
            <div
              class="jsx-4249355495 icon dense"
            >
              <svg
                class="jsx-4069423049 jsx-3997907426 checked"
                viewBox="0 0 14 14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  class="jsx-4069423049 jsx-3997907426 background"
                  height="14"
                  rx="2"
                  width="14"
                  x="0"
                  y="0"
                />
                <path
                  class="jsx-4069423049 jsx-3997907426 border"
                  d="M2,0 L12,0 C13.1045695,0 14,0.8954305 14,2 L14,12 C14,13.1045695 13.1045695,14 12,14 L2,14 C0.8954305,14 0,13.1045695 0,12 L0,2 C0,0.8954305 0.8954305,0 2,0 Z M2,1 C1.44771525,1 1,1.44771525 1,2 L1,12 C1,12.5522847 1.44771525,13 2,13 L12,13 C12.5522847,13 13,12.5522847 13,12 L13,2 C13,1.44771525 12.5522847,1 12,1 L2,1 Z"
                />
                <path
                  class="jsx-4069423049 jsx-3997907426 checkmark"
                  d="M10.3520005,3.30015877 L5.54268293,8.03252726 L3.64634146,6.16652726 L2.28668615,7.5044281 L5.54310367,10.7010587 L11.7145993,4.628307 L10.3520005,3.30015877 Z M5.54268293,9.43547274 L5.61193184,9.36733181 L5.54226219,9.29894127 L5.47309681,9.367 L5.54268293,9.43547274 Z M10.6496329,4.41023392 L10.6495057,4.35652297 L10.5648412,4.35672356 L10.2854007,4.631693 L10.3553166,4.69984123 L10.6496329,4.41023392 Z M3.42891348,7.22439304 L3.35030905,7.22439304 L3.35034135,7.27820863 L3.64634146,7.56947274 L3.71331385,7.5035719 L3.42891348,7.22439304 Z"
                />
                <rect
                  class="jsx-4069423049 jsx-3997907426 indeterminate"
                  height="2"
                  width="8"
                  x="3"
                  y="6"
                />
              </svg>
            </div>
            Use human-readable keys
          </label>
        </div>
      </div>
    </div>
    <div
      class="ui-ModalActions"
    >
      <div
        class="ui-ButtonStrip"
      >
        <div
          class="ui-Button"
        >
          Cancel
        </div>
        <div
          class="ui-Button"
        >
          Download
          <div
            class="jsx-1838964138 btnProgress small"
            data-test="dhis2-uicore-circularloader"
            role="progressbar"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DataDownloadDialogContent renders null when open but layer is null 1`] = `<div />`;
